docs/design/distributed-cache.md

```diff
@@ -0,0 +1,247 @@
+        // Delete item from internal cache implementation
+    }
+
+    private purgeStale() {
+        // Delete any stale entries implementation.
+    }
+
+    clear() {
+        // Clear the cache entirely implementation.
+    }
+}
+```
+
+```javascript
+import { createClient } from 'redis';
+
+class RedisCache implements ICacheClient{
+    /**
+     * The Redis cache used for caching items from requests.
+     *
+     * @private
+     */
+    private readonly cache;
+
+    public constructor() {
+        const client = createClient({
+            url: ''
+        });
+        await client.connect();
+
+        this.cache = client;
+    }
+    get(key: string, callingMethod: string, requestIdPrefix?: string) {
+        // Get item from shared cache implementation
+    }
+
+    set(key: string, value: any, callingMethod: string, ttl?: number, requestIdPrefix?: string) {
+        // Set item to shared cache implementation
+    }
+
+    delete(key: string, callingMethod: string, requestIdPrefix?: string) {
+        // Delete item from shared cache implementation
+    }
+
+    clear() {
+        // Clear the cache entirely implementation.
+    }
+}
+```
+
+```javascript
+class CacheClientService{
+    /**
+     * The LRU cache used for caching items from requests.
+     *
+     * @private
+     */
+    private readonly internalCache;
+
+    /**
+     * The Redis cache used for caching items from requests.
+     *
+     * @private
+     */
+    private readonly sharedCache;
+
+    public constructor() {
+        const internalCache = new LocalLRUCache();
+        const sharedCache = new RedisCache();
+    }
+
+    get(key: string, callingMethod: string, shared: boolean = false, requestIdPrefix?: string) {
+        // Depending on the shared boolean, this method decide from where it should request the data.
+        // Fallbacks to internalCache in case of error from the shared cache.
+        // Getting from shared cache depends on REDIS_ENABLED env. variable
+    }
+
+    set(key: string, value: any, callingMethod: string, shared: boolean = false, ttl?: number, requestIdPrefix?: string) {
+        // Depending on the shared boolean, this method decide where it should save the data.
+        // Fallbacks to internalCache in case of error from the shared cache.
+        // Setting to shared cache depends on REDIS_ENABLED env. variable
+    }
+
+    delete(key: string, callingMethod: string, shared: boolean = false, requestIdPrefix?: string) {
+        // Depending on the shared boolean, this method decide from where it should delete the data.
+        // Fallbacks to internalCache in case of error from the shared cache.
+        // Deleting from shared cache depends on REDIS_ENABLED env. variable
+    }
+
+    clear(shared: boolean = false) {
+        // In case of error does NOT fallback to shared cache.
+        // Clearing from shared cache depends on REDIS_ENABLED env. variable
+    }
+}
+```
+
+### Flowchart
+
+```mermaid
+flowchart TD
```

Comment:
Modification here is that the `Cache Client Service` sits inside each relay instance. Based on the need for share the call will either be serviced in instance by the `LocalLRUCache` or it will go out of the instance to the `RedisCache`.
Just clarifying as the digram may mislead to think the LocalLRUCache is somehow not also in the relay instance.

Reply:
Ah, yes, thank you. Fixed it.